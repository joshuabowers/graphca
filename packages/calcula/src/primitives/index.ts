import { isPrimitive, PrimitiveNode } from '../closures/primitive'
import { Real, real, isReal, EulerMascheroni } from './real'
import { 
  Complex, complex, isComplex, ComplexInfinity, isComplexInfinity 
} from './complex'
import { Boolean, boolean, isBoolean } from './boolean'
import { Nil, nil, isNil } from './nil'
import { NaN, nan, isNaN } from './nan'

export type Primitive = Real | Complex | Boolean | Nil | NaN
export { Real, Complex, Boolean, Nil, NaN }

export type PrimitiveFns = 
| typeof real | typeof complex | typeof boolean

export { real, complex, boolean, nil, nan }
export { isReal, isComplex, isBoolean, isNil, isNaN, isComplexInfinity }
export { isPrimitive, PrimitiveNode }
export { EulerMascheroni, ComplexInfinity }
