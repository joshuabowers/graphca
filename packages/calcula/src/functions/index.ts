import { Unicode } from "../Unicode"
import { 
  Equality, StrictInequality, LessThan, GreaterThan,
  LessThanEquals, GreaterThanEquals,
  equals, notEquals, lessThan, greaterThan,
  lessThanEquals, greaterThanEquals,
  isEquality, isStrictInequality, isLessThan, isGreaterThan,
  isLessThanEquals, isGreaterThanEquals,
  $equals, $notEquals, $lessThan, $greaterThan,
  $lessThanEquals, $greaterThanEquals
} from "./inequality"
import { 
  Complement, Conjunction, Disjunction, ExclusiveDisjunction,
  Implication, AlternativeDenial, JointDenial, Biconditional, ConverseImplication,
  not, and, or, xor, implies, nand, nor, xnor, converse,
  isComplement, isConjunction, isDisjunction, isExclusiveDisjunction,
  isImplication, isAlternativeDenial, isJointDenial, isBiconditional,
  isConverseImplication, isConnective,
  $not, $and, $or, $xor, $implies, $nand, $nor, $xnor, $converse
} from "./connectives"
import { 
  Logarithm, lb, ln, lg, isLogarithm, $log
} from './logarithmic'
import { 
  Cosine, Sine, Tangent, Secant, Cosecant, Cotangent,
  cos, sin, tan, sec, csc, cot,
  isTrigonometric,
  isCosine, isSine, isTangent, isSecant, isCosecant, isCotangent,
  $cos, $sin, $tan, $sec, $csc, $cot
} from './trigonometric'
import { 
  ArcusCosine, ArcusSine, ArcusTangent, 
  ArcusSecant, ArcusCosecant, ArcusCotangent,
  acos, asin, atan, asec, acsc, acot,
  isArcus, isArcusCosine, isArcusSine, isArcusTangent,
  isArcusSecant, isArcusCosecant, isArcusCotangent,
  $acos, $asin, $atan, $asec, $acsc, $acot
} from './arcus'
import { 
  HyperbolicCosine, HyperbolicSine, HyperbolicTangent,
  HyperbolicSecant, HyperbolicCosecant, HyperbolicCotangent,
  cosh, sinh, tanh, sech, csch, coth,
  isHyperbolic, isHyperbolicCosine, isHyperbolicSine, isHyperbolicTangent,
  isHyperbolicSecant, isHyperbolicCosecant, isHyperbolicCotangent,
  $cosh, $sinh, $tanh, $sech, $csch, $coth
} from './hyperbolic'
import { 
  AreaHyperbolicCosine, AreaHyperbolicSine, AreaHyperbolicTangent,
  AreaHyperbolicSecant, AreaHyperbolicCosecant, AreaHyperbolicCotangent,
  acosh, asinh, atanh, asech, acsch, acoth,
  isAreaHyperbolic, isAreaHyperbolicCosine, isAreaHyperbolicSine,
  isAreaHyperbolicTangent, isAreaHyperbolicSecant, isAreaHyperbolicCosecant,
  isAreaHyperbolicCotangent,
  $acosh, $asinh, $atanh, $asech, $acsch, $acoth
} from './areaHyperbolic'
import { 
  Gamma, gamma, isGamma, $gamma
} from './gamma'
import { 
  Absolute, abs, isAbsolute, $abs
} from './absolute'
import { sqrt } from '../arithmetic/exponentiation'

export type InequalityFn = 
| typeof equals | typeof notEquals
| typeof lessThan | typeof greaterThan 
| typeof lessThanEquals | typeof greaterThanEquals

export type ConnectiveFn =
| typeof and | typeof or | typeof xor | typeof implies
| typeof nand | typeof nor | typeof xnor | typeof converse

type Functions = 
| typeof lb | typeof ln | typeof lg
| typeof cos | typeof sin | typeof tan
| typeof sec | typeof csc | typeof cot
| typeof acos | typeof asin | typeof atan
| typeof asec | typeof acsc | typeof acot
| typeof cosh | typeof sinh | typeof tanh
| typeof sech | typeof csch | typeof coth
| typeof acosh | typeof asinh | typeof atanh
| typeof asech | typeof acsch | typeof acoth
| typeof gamma | typeof abs | typeof sqrt

export const inequality = new Map<string, InequalityFn>(
  [
    ['==', equals],
    ['!=', notEquals],
    ['<=', lessThanEquals],
    ['>=', greaterThanEquals],
    ['<', lessThan],
    ['>', greaterThan]
  ]
)

export const connectives = new Map<string, ConnectiveFn>(
  [
    [Unicode.xnor, xnor],
    [Unicode.and, and],
    [Unicode.or, or],
    [Unicode.xor, xor],
    [Unicode.implies, implies],
    [Unicode.nand, nand],
    [Unicode.nor, nor],
    [Unicode.converse, converse],
  ]
)

export const functions = new Map<string, Functions>([
  ['lb', lb],
  ['ln', ln],
  ['lg', lg],
  ['acosh', acosh],
  ['asinh', asinh],
  ['atanh', atanh],
  ['asech', asech],
  ['acsch', acsch],
  ['acoth', acoth],
  ['acos', acos],
  ['asin', asin],
  ['atan', atan],
  ['asec', asec],
  ['acsc', acsc],
  ['acot', acot],
  ['cosh', cosh],
  ['sinh', sinh],
  ['tanh', tanh],
  ['sech', sech],
  ['csch', csch],
  ['coth', coth],
  ['cos', cos],
  ['sin', sin],
  ['tan', tan],
  ['sec', sec],
  ['csc', csc],
  ['cot', cot],
  ['abs', abs],
  [Unicode.gamma, gamma],
  [Unicode.squareRoot, sqrt]
])

export {
  Equality, StrictInequality, LessThan, GreaterThan,
  LessThanEquals, GreaterThanEquals,
  Complement, Conjunction, Disjunction, ExclusiveDisjunction,
  Implication, AlternativeDenial, JointDenial, Biconditional, ConverseImplication, 
  Logarithm, Cosine, Sine, Tangent, Secant, Cosecant, Cotangent,
  ArcusCosine, ArcusSine, ArcusTangent, 
  ArcusSecant, ArcusCosecant, ArcusCotangent,
  HyperbolicCosine, HyperbolicSine, HyperbolicTangent,
  HyperbolicSecant, HyperbolicCosecant, HyperbolicCotangent,
  AreaHyperbolicCosine, AreaHyperbolicSine, AreaHyperbolicTangent,
  AreaHyperbolicSecant, AreaHyperbolicCosecant, AreaHyperbolicCotangent,
  Gamma, Absolute
}

export {
  equals, notEquals, lessThan, greaterThan,
  lessThanEquals, greaterThanEquals,
  not, and, or, xor, implies, nand, nor, xnor, converse,
  lb, ln, lg, cos, sin, tan, sec, csc, cot,
  acos, asin, atan, asec, acsc, acot,
  cosh, sinh, tanh, sech, csch, coth,
  acosh, asinh, atanh, asech, acsch, acoth,
  gamma, abs
}

export {
  isEquality, isStrictInequality, isLessThan, isGreaterThan,
  isLessThanEquals, isGreaterThanEquals,
  isComplement, isConjunction, isDisjunction, isExclusiveDisjunction,
  isImplication, isAlternativeDenial, isJointDenial, isBiconditional,
  isConverseImplication, isConnective, isLogarithm, isTrigonometric,
  isCosine, isSine, isTangent, isSecant, isCosecant, isCotangent,
  isArcus, isArcusCosine, isArcusSine, isArcusTangent,
  isArcusSecant, isArcusCosecant, isArcusCotangent,
  isHyperbolic, isHyperbolicCosine, isHyperbolicSine, isHyperbolicTangent,
  isHyperbolicSecant, isHyperbolicCosecant, isHyperbolicCotangent,
  isAreaHyperbolic, isAreaHyperbolicCosine, isAreaHyperbolicSine,
  isAreaHyperbolicTangent, isAreaHyperbolicSecant, isAreaHyperbolicCosecant,
  isAreaHyperbolicCotangent, isGamma, isAbsolute
}

export {
  $equals, $notEquals, $lessThan, $greaterThan,
  $lessThanEquals, $greaterThanEquals,
  $not, $and, $or, $xor, $implies, $nand, $nor, $xnor, $converse,
  $log, $cos, $sin, $tan, $sec, $csc, $cot,
  $acos, $asin, $atan, $asec, $acsc, $acot,
  $cosh, $sinh, $tanh, $sech, $csch, $coth,
  $acosh, $asinh, $atanh, $asech, $acsch, $acoth,
  $gamma, $abs
}

export { Factorial, factorial, isFactorial, $factorial } from './factorial'
export { 
  Polygamma, polygamma, digamma, isPolygamma, $polygamma 
} from './polygamma'
export { 
  Permutation, Combination, 
  permute, combine, 
  isPermutation, isCombination, 
  $permute, $combine
} from './combinatorics'
