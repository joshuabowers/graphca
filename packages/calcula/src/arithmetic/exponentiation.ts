import { _ } from '@arrows/multimethod'
import { TreeNode, Genera, Species, Notation } from "../utility/tree"
import { real, complex, boolean } from "../primitives"
import { Binary, binary, when, partialRight } from "../closures/binary"
import { deepEquals, isValue } from "../utility/deepEquals"
import { isMultiplication, Multiplication, multiply } from './multiplication'
import { isLogarithm, Logarithm } from '../functions/logarithmic'

export type Exponentiation = Binary<Species.raise, Genera.arithmetic>

export const [raise, isExponentiation, $raise] = binary<Exponentiation>(
  '^', Notation.infix, Species.raise, Genera.arithmetic
)(
  (l, r) => real(l.value.value ** r.value.value),
  (l, r) => {
    const p = Math.hypot(l.value.a, l.value.b), arg = Math.atan2(l.value.b, l.value.a)
    const dLnP = r.value.b * Math.log(p), cArg = r.value.a * arg
    const multiplicand = (p ** r.value.a) * Math.exp(-r.value.b * arg)
    return complex([
      multiplicand * Math.cos(dLnP + cArg),
      multiplicand * Math.sin(dLnP + cArg)
    ])
  },
  (l, r) => boolean(l.value.value || !r.value.value)
)(
  when(
    [isValue(complex([0, 0])), isValue(real(-1))], 
    [complex([Infinity, 0]), 'division by complex zero']
  ),
  when(
    [isValue(real(0)), isValue(real(-1))], 
    [real(Infinity), 'division by zero']
  ),
  when(
    [isValue(real(-0)), isValue(real(-1))], 
    [real(-Infinity), 'division by negative zero']
  ),
  when([isValue(real(0)), _], [real(0), 'powers of 0']),
  when([_, isValue(real(0))], [real(1), 'exponent of 0']),
  when([isValue(real(1)), _], [real(1), 'powers of 1']),
  when([_, isValue(real(1))], (l, _r) => [l, 'exponent of 1']),

  when<TreeNode, Logarithm>(
    (l, r) => isLogarithm(r) && deepEquals(l, r.value.left),
    (_l, r) => [r.value.right, 'inverse function cancellation']
  ),
  when<Exponentiation, TreeNode>(
    (l, _r) => isExponentiation(l),
    (l, r) => [
      raise(l.value.left, multiply(l.value.right, r)), 
      'exponential product'
    ]
  ),
  when<Multiplication, TreeNode>(
    (l, _r) => isMultiplication(l),
    (l, r) => [
      multiply(raise(l.value.left, r), raise(l.value.right, r)), 
      'exponential distribution'
    ]
  )
)

export const reciprocal = partialRight(raise)(real(-1))
export const square = partialRight(raise)(real(2))
export const sqrt = partialRight(raise)(real(0.5))
