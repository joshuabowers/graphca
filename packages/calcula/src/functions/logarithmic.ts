import { Genera, Species, Notation, TreeNode } from "../utility/tree";
import { real, complex, boolean, Complex } from "../primitives";
import { Binary, binary, when, partialLeft } from "../closures/binary";
import { divide } from '../arithmetic'
import { Exponentiation, isExponentiation } from "../arithmetic/exponentiation";
import { deepEquals } from "../utility/deepEquals";

export type Logarithm = Binary<Species.log, Genera.logarithmic>

const lnComplex = (c: Complex) => complex(
  Math.log(Math.hypot(c.raw.a, c.raw.b)),
  Math.atan2(c.raw.b, c.raw.a)
)

export const [log, isLogarithm, $log] = binary<Logarithm>(
  'log', Notation.prefix, Species.log, Genera.logarithmic
)(
  (l, r) => real(Math.log(r.value.raw) / Math.log(l.value.raw)), 
  (l, r) => {
    const n = lnComplex(r.value)
    if(l.value.raw.a === Math.E && l.value.raw.b === 0){ return n }
    return divide(lnComplex(r.value), lnComplex(l.value))
  },
  (l, r) => boolean(l.value.raw || !r.value.raw)
)(
  when<TreeNode, Exponentiation>(
    (l, r) => isExponentiation(r) && deepEquals(l, r.value.left),
    (_l, r) => [r.value.right, 'inverse operation cancellation']
  )
)

const fromLog = partialLeft(log)
export const lb = fromLog(real(2))
export const ln = fromLog(real(Math.E))
export const lg = fromLog(real(10))
