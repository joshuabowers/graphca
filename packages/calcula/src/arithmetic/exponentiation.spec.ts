import { unit } from '../monads/writer'
import { Clades, Genera, Species } from '../utility/tree'
import { real, complex } from '../primitives'
import { variable } from '../variable'
import { multiply } from './multiplication'
import { raise, reciprocal, square, sqrt, $raise } from './exponentiation'
import { log, lb, ln, lg } from '../functions/logarithmic'
import { 
  expectToEqualWithSnapshot, expectCloseTo 
} from '../utility/expectations'

describe('$raise', () => {
  it('generates an Exponentiation for a pair of TreeNode inputs', () => {
    expect(
      $raise(unit(variable('x').value), unit(variable('y').value))[0]
    ).toEqual({
      clade: Clades.binary, genus: Genera.arithmetic, species: Species.raise,
      left: unit(variable('x').value), right: unit(variable('y').value)
    })
  })
})

describe('raise', () => {
  it('computes the value of a real raised to a real', () => {
    expectToEqualWithSnapshot(
      raise(real(2), real(3)),
      real(8)
    )
  })

  it('calculates the value of raising one complex to another', () => {
    expectCloseTo(raise(complex(0, 1), complex(0, 1)), complex(0.20787957635076, 0), 10)
  })

  it('calculates the value of raising a real to a complex number', () => {
    expectCloseTo(raise(real(-2), complex(3, 4)), complex(0.000026020793185, -0.000010062701000), 10)
  })

  it('calculates the value of raising a complex number to a real', () => {
    expectCloseTo(raise(complex(2, 3), real(5)), complex(122, -597), 10)
  })

  it('calculates the complex square root', () => {
    expectCloseTo(
      raise(complex(2, 3), real(0.5)), 
      complex(1.67414922803554, 0.89597747612983), 
      10
    )
  })

  it('returns 0 when the base is 0', () => {
    expectToEqualWithSnapshot(
      raise(real(0), variable('x')),
      real(0)
    )
  })

  it('returns 1 when the exponent is 0', () => {
    expectToEqualWithSnapshot(
      raise(variable('x'), real(0)),
      real(1)
    )
  })

  it('returns 1 when the base is 1', () => {
    expectToEqualWithSnapshot(
      raise(real(1), variable('x')),
      real(1)
    )
  })

  it('returns the base whenever the exponent is 1', () => {
    expectToEqualWithSnapshot(
      raise(variable('x'), real(1)),
      variable('x')
    )
  })

  it('returns the sub-expression of an lb if base 2', () => {
    expectToEqualWithSnapshot(
      raise(real(2), lb(variable('x'))),
      variable('x')
    )
  })

  it('returns the sub-expression of an ln if base e', () => {
    expectToEqualWithSnapshot(
      raise(real(Math.E), ln(variable('x'))),
      variable('x')
    )
  })

  it('returns the sub-expression of an lg if base 10', () => {
    expectToEqualWithSnapshot(
      raise(real(10), lg(variable('x'))),
      variable('x')
    )
  })
  
  it('returns the value of a logarithm if raising similar base to it', () => {
    expectToEqualWithSnapshot(
      raise(complex(0, 1), log(complex(0, 1), variable('x'))),
      variable('x')
    )
  })

  it('multiplies the exponent of a base exponential against the exponent', () => {
    expectToEqualWithSnapshot(
      raise(raise(variable('x'), variable('y')), variable('z')),
      raise(variable('x'), multiply(variable('y'), variable('z')))
    )
  })

  it('converts a base multiplication into a product of exponentiations', () => {
    expectToEqualWithSnapshot(
      raise(multiply(variable('x'), variable('y')), variable('z')),
      multiply(
        raise(variable('x'), variable('z')),
        raise(variable('y'), variable('z'))
      )
    )
  })
  
  it('creates an Exponentiation when given non-constants', () => {
    expectToEqualWithSnapshot(
      raise(variable('x'), real(3)),
      $raise(variable('x'), real(3))[0]
    )
  })
})

describe('reciprocal', () => {
  it('raises its argument to the power of -1', () => {
    expectToEqualWithSnapshot(
      reciprocal(variable('x')),
      raise(variable('x'), real(-1))
    )
  })

  it('raises complex 1 to -1 correctly', () => {
    expectToEqualWithSnapshot(
      reciprocal(complex(1, 0)),
      complex(1, 0)
    )
  })

  it('calculates a complex reciprocal correctly', () => {
    expectCloseTo(reciprocal(complex(1, 1)), complex(0.5, -0.5), 10)
  })

  it('calculates a reciprocal of complex 0 correctly', () => {
    expectCloseTo(reciprocal(complex(0, 0)), complex(Infinity, 0), 10)
  })
})

describe('square', () => {
  it('raises its argument to the power of 2', () => {
    expectToEqualWithSnapshot(
      square(variable('x')),
      raise(variable('x'), real(2))
    )
  })

  it('calculates the proper square of a complex number', () => {
    expectCloseTo(square(complex(1, 1)), complex(0, 2), 10)
  })
})

describe('sqrt', () => {
  it('raises its argument to the power of 0.5', () => {
    expectToEqualWithSnapshot(
      sqrt(variable('x')),
      raise(variable('x'), real(0.5))
    )
  })

  it('calculates the proper sqrt of a complex number', () => {
    expectCloseTo(sqrt(complex(1, 1)), complex(1.098684113467, 0.455089860562), 10)
  })

  it('calculates the complex sqrt of a complex number', () => {
    expectCloseTo(sqrt(complex(3, 1)), complex(1.755317301824, 0.284848784593), 10)
  })
})
