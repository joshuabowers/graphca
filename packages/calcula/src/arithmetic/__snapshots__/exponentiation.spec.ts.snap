// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`raise computes the value of a real raised to a real 1`] = `
[
  [
    "(2^3)",
    "identified exponentiation",
  ],
  [
    "([2]^3)",
    "processing left operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "({2}^[3])",
    "processed left operand; processing right operand",
  ],
  [
    "3",
    "created real",
  ],
  [
    "(2^{3})",
    "processed right operand",
  ],
  [
    "(2^3)",
    "real exponentiation",
  ],
  [
    "8",
    "created real",
  ],
]
`;

exports[`raise converts a base multiplication into a product of exponentiations 1`] = `
[
  [
    "((x*y)^z)",
    "identified exponentiation",
  ],
  [
    "([(x*y)]^z)",
    "processing left operand",
  ],
  [
    "(x*y)",
    "identified multiplication",
  ],
  [
    "([x]*y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}*[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x*{y})",
    "processed right operand",
  ],
  [
    "(x*y)",
    "created multiplication",
  ],
  [
    "({(x*y)}^[z])",
    "processed left operand; processing right operand",
  ],
  [
    "z",
    "referenced variable",
  ],
  [
    "((x*y)^{z})",
    "processed right operand",
  ],
  [
    "((x*y)^z)",
    "exponential distribution",
  ],
  [
    "((x^z)*(y^z))",
    "identified multiplication",
  ],
  [
    "([(x^z)]*(y^z))",
    "processing left operand",
  ],
  [
    "(x^z)",
    "identified exponentiation",
  ],
  [
    "([x]^z)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}^[z])",
    "processed left operand; processing right operand",
  ],
  [
    "z",
    "referenced variable",
  ],
  [
    "(x^{z})",
    "processed right operand",
  ],
  [
    "(x^z)",
    "created exponentiation",
  ],
  [
    "({(x^z)}*[(y^z)])",
    "processed left operand; processing right operand",
  ],
  [
    "(y^z)",
    "identified exponentiation",
  ],
  [
    "([y]^z)",
    "processing left operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "({y}^[z])",
    "processed left operand; processing right operand",
  ],
  [
    "z",
    "referenced variable",
  ],
  [
    "(y^{z})",
    "processed right operand",
  ],
  [
    "(y^z)",
    "created exponentiation",
  ],
  [
    "((x^z)*{(y^z)})",
    "processed right operand",
  ],
  [
    "((x^z)*(y^z))",
    "created multiplication",
  ],
]
`;

exports[`raise creates an Exponentiation when given non-constants 1`] = `
[
  [
    "(x^3)",
    "identified exponentiation",
  ],
  [
    "([x]^3)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}^[3])",
    "processed left operand; processing right operand",
  ],
  [
    "3",
    "created real",
  ],
  [
    "(x^{3})",
    "processed right operand",
  ],
  [
    "(x^3)",
    "created exponentiation",
  ],
]
`;

exports[`raise multiplies the exponent of a base exponential against the exponent 1`] = `
[
  [
    "((x^y)^z)",
    "identified exponentiation",
  ],
  [
    "([(x^y)]^z)",
    "processing left operand",
  ],
  [
    "(x^y)",
    "identified exponentiation",
  ],
  [
    "([x]^y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}^[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x^{y})",
    "processed right operand",
  ],
  [
    "(x^y)",
    "created exponentiation",
  ],
  [
    "({(x^y)}^[z])",
    "processed left operand; processing right operand",
  ],
  [
    "z",
    "referenced variable",
  ],
  [
    "((x^y)^{z})",
    "processed right operand",
  ],
  [
    "((x^y)^z)",
    "exponential product",
  ],
  [
    "(x^(y*z))",
    "identified exponentiation",
  ],
  [
    "([x]^(y*z))",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}^[(y*z)])",
    "processed left operand; processing right operand",
  ],
  [
    "(y*z)",
    "identified multiplication",
  ],
  [
    "([y]*z)",
    "processing left operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "({y}*[z])",
    "processed left operand; processing right operand",
  ],
  [
    "z",
    "referenced variable",
  ],
  [
    "(y*{z})",
    "processed right operand",
  ],
  [
    "(y*z)",
    "created multiplication",
  ],
  [
    "(x^{(y*z)})",
    "processed right operand",
  ],
  [
    "(x^(y*z))",
    "created exponentiation",
  ],
]
`;

exports[`raise returns 0 when the base is 0 1`] = `
[
  [
    "(0^x)",
    "identified exponentiation",
  ],
  [
    "([0]^x)",
    "processing left operand",
  ],
  [
    "0",
    "created real",
  ],
  [
    "({0}^[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(0^{x})",
    "processed right operand",
  ],
  [
    "(0^x)",
    "powers of 0",
  ],
  [
    "0",
    "created real",
  ],
]
`;

exports[`raise returns 1 when the base is 1 1`] = `
[
  [
    "(1^x)",
    "identified exponentiation",
  ],
  [
    "([1]^x)",
    "processing left operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "({1}^[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(1^{x})",
    "processed right operand",
  ],
  [
    "(1^x)",
    "powers of 1",
  ],
  [
    "1",
    "created real",
  ],
]
`;

exports[`raise returns 1 when the exponent is 0 1`] = `
[
  [
    "(x^0)",
    "identified exponentiation",
  ],
  [
    "([x]^0)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}^[0])",
    "processed left operand; processing right operand",
  ],
  [
    "0",
    "created real",
  ],
  [
    "(x^{0})",
    "processed right operand",
  ],
  [
    "(x^0)",
    "exponent of 0",
  ],
  [
    "1",
    "created real",
  ],
]
`;

exports[`raise returns the base whenever the exponent is 1 1`] = `
[
  [
    "(x^1)",
    "identified exponentiation",
  ],
  [
    "([x]^1)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}^[1])",
    "processed left operand; processing right operand",
  ],
  [
    "1",
    "created real",
  ],
  [
    "(x^{1})",
    "processed right operand",
  ],
  [
    "(x^1)",
    "exponent of 1",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`raise returns the sub-expression of an lb if base 2 1`] = `
[
  [
    "(2^log(2,x))",
    "identified exponentiation",
  ],
  [
    "([2]^log(2,x))",
    "processing left operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "({2}^[log(2,x)])",
    "processed left operand; processing right operand",
  ],
  [
    "log(2,x)",
    "identified logarithm",
  ],
  [
    "log([2],x)",
    "processing left operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "log({2},[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "log(2,{x})",
    "processed right operand",
  ],
  [
    "log(2,x)",
    "created logarithm",
  ],
  [
    "(2^{log(2,x)})",
    "processed right operand",
  ],
  [
    "(2^log(2,x))",
    "inverse function cancellation",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`raise returns the sub-expression of an lg if base 10 1`] = `
[
  [
    "(10^log(10,x))",
    "identified exponentiation",
  ],
  [
    "([10]^log(10,x))",
    "processing left operand",
  ],
  [
    "10",
    "created real",
  ],
  [
    "({10}^[log(10,x)])",
    "processed left operand; processing right operand",
  ],
  [
    "log(10,x)",
    "identified logarithm",
  ],
  [
    "log([10],x)",
    "processing left operand",
  ],
  [
    "10",
    "created real",
  ],
  [
    "log({10},[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "log(10,{x})",
    "processed right operand",
  ],
  [
    "log(10,x)",
    "created logarithm",
  ],
  [
    "(10^{log(10,x)})",
    "processed right operand",
  ],
  [
    "(10^log(10,x))",
    "inverse function cancellation",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`raise returns the sub-expression of an ln if base e 1`] = `
[
  [
    "(𝒆^log(𝒆,x))",
    "identified exponentiation",
  ],
  [
    "([𝒆]^log(𝒆,x))",
    "processing left operand",
  ],
  [
    "𝒆",
    "created real",
  ],
  [
    "({𝒆}^[log(𝒆,x)])",
    "processed left operand; processing right operand",
  ],
  [
    "log(𝒆,x)",
    "identified logarithm",
  ],
  [
    "log([𝒆],x)",
    "processing left operand",
  ],
  [
    "𝒆",
    "created real",
  ],
  [
    "log({𝒆},[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "log(𝒆,{x})",
    "processed right operand",
  ],
  [
    "log(𝒆,x)",
    "created logarithm",
  ],
  [
    "(𝒆^{log(𝒆,x)})",
    "processed right operand",
  ],
  [
    "(𝒆^log(𝒆,x))",
    "inverse function cancellation",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`raise returns the value of a logarithm if raising similar base to it 1`] = `
[
  [
    "(0+1𝒊^log(0+1𝒊,x))",
    "identified exponentiation",
  ],
  [
    "([0+1𝒊]^log(0+1𝒊,x))",
    "processing left operand",
  ],
  [
    "0+1𝒊",
    "created complex",
  ],
  [
    "({0+1𝒊}^[log(0+1𝒊,x)])",
    "processed left operand; processing right operand",
  ],
  [
    "log(0+1𝒊,x)",
    "identified logarithm",
  ],
  [
    "log([0+1𝒊],x)",
    "processing left operand",
  ],
  [
    "0+1𝒊",
    "created complex",
  ],
  [
    "log({0+1𝒊},[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "log(0+1𝒊,{x})",
    "processed right operand",
  ],
  [
    "log(0+1𝒊,x)",
    "created logarithm",
  ],
  [
    "(0+1𝒊^{log(0+1𝒊,x)})",
    "processed right operand",
  ],
  [
    "(0+1𝒊^log(0+1𝒊,x))",
    "inverse function cancellation",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`reciprocal raises complex 1 to -1 correctly 1`] = `
[
  [
    "(1+0𝒊^-1)",
    "identified exponentiation",
  ],
  [
    "([1+0𝒊]^-1)",
    "processing left operand",
  ],
  [
    "1+0𝒊",
    "created complex",
  ],
  [
    "({1+0𝒊}^[-1])",
    "processed left operand; processing right operand",
  ],
  [
    "-1",
    "created real",
  ],
  [
    "-1+0𝒊",
    "cast to Complex from Real",
  ],
  [
    "(1+0𝒊^{-1+0𝒊})",
    "processed right operand",
  ],
  [
    "(1+0𝒊^-1+0𝒊)",
    "complex exponentiation",
  ],
  [
    "1+0𝒊",
    "created complex",
  ],
]
`;

exports[`reciprocal raises its argument to the power of -1 1`] = `
[
  [
    "(x^-1)",
    "identified exponentiation",
  ],
  [
    "([x]^-1)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}^[-1])",
    "processed left operand; processing right operand",
  ],
  [
    "-1",
    "created real",
  ],
  [
    "(x^{-1})",
    "processed right operand",
  ],
  [
    "(x^-1)",
    "created exponentiation",
  ],
]
`;

exports[`sqrt raises its argument to the power of 0.5 1`] = `
[
  [
    "(x^0.5)",
    "identified exponentiation",
  ],
  [
    "([x]^0.5)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}^[0.5])",
    "processed left operand; processing right operand",
  ],
  [
    "0.5",
    "created real",
  ],
  [
    "(x^{0.5})",
    "processed right operand",
  ],
  [
    "(x^0.5)",
    "created exponentiation",
  ],
]
`;

exports[`square raises its argument to the power of 2 1`] = `
[
  [
    "(x^2)",
    "identified exponentiation",
  ],
  [
    "([x]^2)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}^[2])",
    "processed left operand; processing right operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "(x^{2})",
    "processed right operand",
  ],
  [
    "(x^2)",
    "created exponentiation",
  ],
]
`;
